Swiftless OpenStack: store bootstrap ignition configs in Glance

Fedosin opened the pull request:
Now we store bootstrap ignition configs in Swift storage, but unfortunately Swift is not available on all OpenStack clouds. It leads to the fact that we can't deploy OpenShift on such swiftless clouds.

This PR allows to store configs in Glance, which is available on 100% of OpenStack clouds.

Fedosin commented:
openstack tests will fail until we upgrade the rhcos image in the installer

Fedosin commented:
/retest

Fedosin commented:
/retest

Fedosin commented:
/retest

Fedosin commented:
/test e2e-aws

pierreprinetti commented:
/retest

Fedosin commented:
/retest

wking commented:
Looks like [the OpenStack job][1] failed to come up?  [AWS][2] came up fine, so the bump to Ignition spec 2.4 doesn't seem to have caused a problem.  Still would be nice to see those green though.

/retest

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/2960/pull-ci-openshift-installer-master-e2e-openstack/3765
[2]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/2960/pull-ci-openshift-installer-master-e2e-aws/9742

abhinavdahiya commented:
Dropping the Swift from destroy means that if I created a cluster using 4.2 and tried to destroy the closer with this new installer, we leak the Swift objects

iamemilio commented:
That's a good point. I think we have to support swift deletion indefinitely.

iamemilio commented:
Once you address Abhinav's comment, LGTM

wking commented:
Hmm, [fresh OpenStack run][1] died the same way?

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/2960/pull-ci-openshift-installer-master-e2e-openstack/3778

Fedosin commented:
yeah,to make OpenStack test pass we need to bump Ignition in the pre-release image first: https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/pre-release/
Because the minimal required Ignition version for this change is 0.35.
For instance, the latest 4.3 release already has it: https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.3/latest/

Fedosin commented:
@abhinavdahiya I didn't drop Swift from the destroy module. I just added a check that if there is no Swift service we are fine with it: https://github.com/openshift/installer/pull/2960/files#diff-3420abf882f61be73bc30ef12460687dR547-R550 `deleteContainers` function in `pkg/destroy/openstack/openstack.go`
So, cluster destruction will work on all OCP versions after this change.

I removed Swift code from bootstrap destruction (`pkg/destroy/openstack/swift.go`), and I think this is right. Now we create a Swift container, upload the config there, give the link to the bootstrap machine and then immediately destroy the container when bootstrapping is over using `DeleteSwiftContainer` function from `pkg/destroy/openstack/swift.go`.
With this patch instead of creating a container, we do the same things with Glance image.

So, this patch only affects the bootstrapping process and all other parts stay the same.

wking commented on pkg/destroy/openstack/swift.go:
```diff
@@ -1,43 +0,0 @@
-)
-
-// DeleteSwiftContainer deletes a container and all of its objects.
-func DeleteSwiftContainer(name string, cloud string) error {
```
It is strange that you can now safely remove this.  What if folks want to delete a cluster created by a previous installer (and so using Swift) which got a Swift bootstrap Ignition store, but then failed to install.  Wouldn't that leak the Swift container with `destroy cluster` not using this function?  Of course, the fact that nothing else in `pkg/destroy/openstack` calls this function means that leak is not *new* in this PR.  But it still seems like a leak to me.

wking replied:
Never mind, @Fedosin points out that this function was pretty similar to [`deleteContainers`][1], which is left alone by this PR.

[1]: https://github.com/openshift/installer/blob/96d416ee36d8660c3f2cb33863587455278f4ed2/pkg/destroy/openstack/openstack.go#L541

pierreprinetti commented:
/retest

mandre commented:
/lgtm
/hold until we bump the rhcos image https://github.com/openshift/installer/pull/2971

Fedosin commented:
openstack tests are green with the RHCOS bump https://github.com/openshift/installer/pull/2971: https://github.com/openshift/installer/pull/2973

sdodson commented:
/approve

Fedosin commented:
/test e2e-openstack

Fedosin commented:
openstack tests are green

Fedosin commented:
/hold cancel
